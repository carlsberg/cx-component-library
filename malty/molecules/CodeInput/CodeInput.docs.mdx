---
labels: ['react', 'typescript', 'codeinput']
description: 'A Code Input component.'
---
