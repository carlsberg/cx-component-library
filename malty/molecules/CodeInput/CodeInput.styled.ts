import { StyledIcon } from '@carlsberggroup/malty.atoms.icon';
import styled, { css } from 'styled-components';

export const StyledCodeInputContainer = styled.div`
  display: flex;
  flex-direction: column;
  font-family: ${({ theme }) => theme.typography.global['font-family'].value};
`;

export const StyledLabel = styled.label`
  color: ${({ theme }) => theme.color.default.value};
  font-size: ${({ theme }) => theme.typography.text.small['font-size'].value}px;
  padding-bottom: ${({ theme }) => theme.variables.codeInput.paddingBottom.value}px;
  font-weight: bold;
`;

export const StyledError = styled.label`
  color: ${({ theme }) => theme.color.system.failStrong.value};
  font-size: ${({ theme }) => theme.typography.text.small['font-size'].value}px;
  font-weight: bold;
  padding-top: ${({ theme }) => theme.variables.codeInput.paddingTop.value}px;
`;

export const StyledCodeInputWrapper = styled.div<{
  isIconLeft?: boolean;
}>`
  position: relative;
  display: flex;
  ${StyledIcon} {
    position: absolute;
    top: ${({ theme }) => theme.variables.codeInput.icontTop.value}%;
    transform: translateY(${({ theme }) => theme.variables.codeInput.iconTopTransform.value}%);
    ${({ isIconLeft }) => css`
      ${isIconLeft ? 'left' : 'right'}: ${({ theme }) => theme.variables.codeInput.iconSmallPadding.value}px;
    `}
  }
`;

export const StyledCodeInput = styled.input<{
  disabled?: boolean;
  sizing?: string;
  hasIcon?: boolean;
  isIconLeft?: boolean;
}>`
  flex: 1 1 auto;
  font-weight: bold;
  font-size: ${({ theme }) => theme.typography.text['medium-small']['font-size'].value}px;
  transition: 0.25s ease-in-out;
  transition-property: border-color, color;
  border: 1px solid ${({ theme }) => theme.color.form.calendarAvailable.value};
  color: ${({ theme }) => theme.color.information.indirect.value};
  height: ${({ sizing }) => sizing}px;
  ::placeholder {
    opacity: 0.8;
    color: ${({ theme }) => theme.color.information.indirect.value};
  }

  ${({ theme, hasIcon, isIconLeft }) => {
    const rightPadding = isIconLeft
      ? `${theme.variables.codeInput.iconLargePadding.value}px`
      : `${theme.variables.codeInput.iconSmallPadding.value}px`;
    const leftPadding = isIconLeft
      ? `${theme.variables.codeInput.iconSmallPadding.value}px`
      : `${theme.variables.codeInput.iconLargePadding.value}px`;
    return hasIcon
      ? css`
          padding: 0 ${leftPadding} 0 ${rightPadding};
        `
      : css`
          padding: 0 ${theme.variables.codeInput.iconSmallPadding.value}px;
        `;
  }}

  ${({ disabled }) =>
    disabled
      ? css`
          background-color: ${({ theme }) => theme.color.form.formSelect.value};
        `
      : css`
          &:hover,
          &:focus {
            outline: none;
          }
          &:hover {
            border-color: ${({ theme }) => theme.color.information.indirect.value};
          }
          &:focus {
            border-color: ${({ theme }) => theme.color.form.calendarSpecial.value};
            color: ${({ theme }) => theme.color.form.calendarSpecial.value};
          }
        `}
`;
