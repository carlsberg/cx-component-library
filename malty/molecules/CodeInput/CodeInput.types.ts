import { IconNamesTypes } from '@carlsberg/malty.atoms.icon';

export interface CodeInputProps extends React.HTMLAttributes<HTMLInputElement> {
  value: string;
  label?: string;
  onValueChange: (value: string) => void;
  type: CodeInputType;
  placeholder?: string;
  error?: string;
  icon?: IconNamesTypes;
  isIconLeft?: boolean;
  isDisabled?: boolean;
  size?: SizeTypes;
}

export enum CodeInputType {
  Text = 'text',
  Number = 'number',
  Email = 'email',
  Password = 'password',
  Date = 'date',
  Search = 'search',
  Telephone = 'tel',
  URL = 'url'
}

export enum SizeTypes {
  Medium = 'Medium',
  Large = 'Large'
}

export enum Sizes {
  Medium = 48,
  Large = 58
}
