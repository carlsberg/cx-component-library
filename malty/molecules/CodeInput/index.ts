export { CodeInput } from './CodeInput';
export type { CodeInputProps } from './CodeInput.types';
