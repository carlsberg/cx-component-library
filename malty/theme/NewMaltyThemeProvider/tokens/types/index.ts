export type { PrimitiveBorderRadiusTokens, SemanticBorderRadiusTokens } from './borderRadius.types';
export type { PrimitiveBorderTokens, SemanticBorderTokens } from './borders.types';
export type { PrimitiveColorsTokens, SemanticColorsTokens } from './colors.types';
export type { GridTokens } from './grid.types';
export type { OpacityTokens } from './opacity.types';
export type { ShadowTokens } from './shadows.types';
export type {
  AdditionalPrimitives,
  BorderRadiusTokens,
  ColorTokens,
  SemanticOverrides,
  SemanticTokens
} from './shared.types';
export type { SizeTokens } from './sizes.types';
export type { TypographyTokens } from './typography.types';
