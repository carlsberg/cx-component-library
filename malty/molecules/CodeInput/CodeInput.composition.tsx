import { MaltyThemeProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useState } from 'react';
import { CodeInput } from './CodeInput';
import { CodeInputType } from './CodeInput.types';

export const BasicCodeInput = () => {
  const [value, setValue] = useState('');
  return (
    <MaltyThemeProvider theme="global">
      <CodeInput type={CodeInputType.Text} value={value} onValueChange={(newValue: string) => setValue(newValue)} />
    </MaltyThemeProvider>
  );
};
