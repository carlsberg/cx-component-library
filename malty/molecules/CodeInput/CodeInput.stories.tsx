import { IconNamesTypes } from '@carlsberg/malty.atoms.icon';
import { Meta, Story } from '@storybook/react';
import React, { useState } from 'react';
import { CodeInput } from './CodeInput';
import { CodeInputProps, CodeInputType, SizeTypes } from './CodeInput.types';

export default {
  title: 'Molecules/CodeInput',
  component: CodeInput,
  argTypes: {
    label: { control: 'text' },
    placeholder: { control: 'text' },
    error: { control: 'text' },
    size: {
      options: Object.values(SizeTypes),
      control: {
        type: 'radio'
      }
    },
    type: {
      options: Object.values(CodeInputType),
      control: {
        type: 'select'
      }
    },
    icon: {
      options: Object.values(IconNamesTypes),
      control: {
        type: 'select'
      }
    },
    isDisabled: { control: 'boolean' },
    isIconLeft: { control: 'boolean' },
    value: {
      table: {
        disable: true
      }
    },
    onValueChange: {
      table: {
        disable: true
      }
    }
  }
} as Meta;

const Template: Story<CodeInputProps> = ({
  value,
  size,
  label,
  type,
  icon,
  placeholder,
  error,
  isDisabled,
  isIconLeft
}: CodeInputProps) => {
  const [stateValue, setStateValue] = useState(value);
  return (
    <CodeInput
      size={size}
      label={label}
      type={type}
      icon={icon}
      placeholder={placeholder}
      error={error}
      isDisabled={!!isDisabled}
      value={stateValue}
      isIconLeft={isIconLeft}
      onValueChange={(newValue: string) => setStateValue(newValue)}
    />
  );
};

export const Main = Template.bind({});
Main.args = {
  size: SizeTypes.Large,
  label: 'Label',
  type: CodeInputType.Text,
  icon: IconNamesTypes.Search,
  placeholder: 'Placeholder',
  error: 'Error text',
  isDisabled: false,
  isIconLeft: false
};
