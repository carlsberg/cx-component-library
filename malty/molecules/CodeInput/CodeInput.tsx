import { Icon, IconColors as Colors, IconSizesTypes as IconSizes } from '@carlsberggroup/malty.atoms.icon';
import React, { useMemo } from 'react';
import { v4 as uuid } from 'uuid';
import {
  StyledCodeInput,
  StyledCodeInputContainer,
  StyledCodeInputWrapper,
  StyledError,
  StyledLabel
} from './CodeInput.styled';
import { CodeInputProps, Sizes, SizeTypes } from './CodeInput.types';

export const CodeInput = ({
  value,
  onValueChange,
  label,
  type,
  placeholder,
  error,
  icon,
  isIconLeft,
  isDisabled,
  size = SizeTypes.Medium
}: CodeInputProps) => {
  const id = useMemo(() => uuid(), []);
  return (
    <StyledCodeInputContainer>
      <StyledLabel htmlFor={id}>{label}</StyledLabel>
      <StyledCodeInputWrapper isIconLeft={isIconLeft}>
        <StyledCodeInput
          name={id}
          id={id}
          value={value}
          placeholder={placeholder}
          disabled={isDisabled}
          size={Sizes[size]}
          hasIcon={!!icon}
          isIconLeft={isIconLeft}
          onChange={(e) => onValueChange((e.target as HTMLInputElement).value)}
          type={type}
        />
        {icon && <Icon name={icon} color={Colors.Primary} size={IconSizes.Medium} />}
      </StyledCodeInputWrapper>
      {error && <StyledError>{error}</StyledError>}
    </StyledCodeInputContainer>
  );
};
