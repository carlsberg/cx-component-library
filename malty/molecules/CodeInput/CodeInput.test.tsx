import { jsonRenderer, render, screen } from '@carlsberg/malty.utils.test';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { CodeInput } from './CodeInput';
import { CodeInputType } from './CodeInput.types';

jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));

const mockFn = jest.fn();

describe('input', () => {
  it('matches snapshot', () => {
    const view = jsonRenderer(<CodeInput value="Value" onValueChange={mockFn} type={CodeInputType.Text} />);
    expect(view).toMatchSnapshot();
  });

  it('renders elements', () => {
    render(
      <CodeInput
        value="Value text"
        label="Label text"
        placeholder="Placeholder text"
        onValueChange={mockFn}
        type={CodeInputType.Text}
        error="Error text"
      />
    );
    expect(screen.getByLabelText('Label text')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Placeholder text')).toBeInTheDocument();
    expect(screen.getByText('Error text')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Value text')).toBeInTheDocument();
  });

  it('calls onValueChange when typing', () => {
    const onValueChange = jest.fn();
    const { rerender } = render(
      <CodeInput
        value="Initial value"
        label="CodeInput label"
        onValueChange={onValueChange}
        type={CodeInputType.Text}
      />
    );
    const input = screen.getByDisplayValue('Initial value');
    userEvent.type(input, 'Test');
    expect(onValueChange).toHaveBeenCalledTimes(4);

    rerender(
      <CodeInput value="Test" label="CodeInput label" onValueChange={onValueChange} type={CodeInputType.Text} />
    );
    expect(screen.getByDisplayValue('Test')).toBeInTheDocument();
  });
});
