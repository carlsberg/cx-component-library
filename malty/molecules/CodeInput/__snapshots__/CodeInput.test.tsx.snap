// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input matches snapshot 1`] = `
<div
  className="sc-gtsrHT jYcUgd"
>
  <label
    className="sc-dlnjwi dbGmPU"
    htmlFor="00000000-0000-0000-0000-000000000000"
  />
  <div
    className="sc-eCApnc eRkmYH"
  >
    <input
      className="sc-jSFjdj ghvDJB"
      id="00000000-0000-0000-0000-000000000000"
      name="00000000-0000-0000-0000-000000000000"
      onChange={[Function]}
      type="text"
      value="Value"
    />
  </div>
</div>
`;
